File: packages/lynx-core/ios/SigxFontScale.swift
import UIKit

/// App-level policy for how the OS text-size setting maps onto the Lynx
/// engine's font scale. Stamped from `signalx.config.ts` (`fontScale: {…}`)
/// by the managed `LynxSetupService.initialize()` before any LynxView exists.
struct SigxFontScalePolicy {
    /// Follow the OS setting. `false` pins the effective scale to 1.0.
    var follow: Bool = true
    /// Lower clamp on the applied scale.
    var min: CGFloat = 0.5
    /// Upper clamp on the applied scale.
    var max: CGFloat = 2.0
}

/// Derives the effective Lynx `fontScale` from Dynamic Type.
///
/// The OS value is read through `UIFontMetrics(.body)` rather than a content-
/// size-category table so the scale tracks Apple's real (non-linear) curve,
/// including the accessibility sizes (~3.1× at AX-XXXL).
enum SigxFontScale {
    /// Set once at startup by the managed `LynxSetupService`; read by the
    /// host's LynxView builder and by `FontScalePublisher`.
    static var policy = SigxFontScalePolicy()

    /// Raw OS scale (1.0 at the default "Large" setting), unclamped.
    static func osScale(for traits: UITraitCollection? = nil) -> CGFloat {
        let metrics = UIFontMetrics(forTextStyle: .body)
        // 17pt is the .body base size at the default content size category.
        let scaled = traits.map { metrics.scaledValue(for: 17, compatibleWith: $0) }
            ?? metrics.scaledValue(for: 17)
        // Round to 3 decimals so dedupe comparisons are stable.
        return (scaled / 17 * 1000).rounded() / 1000
    }

    /// OS scale with the app policy applied — what the engine receives.
    static func effectiveScale(for traits: UITraitCollection? = nil) -> CGFloat {
        guard policy.follow else { return 1.0 }
        return Swift.min(Swift.max(osScale(for: traits), policy.min), policy.max)
    }
}

File: packages/lynx-core/android/com/sigx/core/SigxFontScale.kt
package com.sigx.core

import android.content.Context
import android.content.res.Configuration

/**
 * App-level policy for how the OS font-size setting maps onto the Lynx
 * engine's font scale. Stamped from `signalx.config.ts` (`fontScale: {…}`)
 * by the managed `MainActivity.onCreate` before any LynxView is built.
 */
data class SigxFontScalePolicy(
    /** Follow the OS setting. `false` pins the effective scale to 1.0. */
    val follow: Boolean = true,
    /** Lower clamp on the applied scale. */
    val min: Float = 0.5f,
    /** Upper clamp on the applied scale. */
    val max: Float = 2.0f,
)

/**
 * Derives the effective Lynx `fontScale` from the system
 * [Configuration.fontScale] with the app policy applied. Read by the host's
 * LynxViewBuilder seeds and by [FontScalePublisher].
 */
object SigxFontScale {
    @JvmStatic
    var policy = SigxFontScalePolicy()

    /** Raw OS scale (1.0 at the default setting), unclamped. */
    @JvmStatic
    fun osScale(context: Context): Float = context.resources.configuration.fontScale

    /** OS scale with the app policy applied — what the engine receives. */
    @JvmStatic
    fun effectiveScale(context: Context): Float =
        effective(context.resources.configuration)

    internal fun effective(config: Configuration): Float =
        if (!policy.follow) 1.0f else config.fontScale.coerceIn(policy.min, policy.max)
}
